File: java/leetcode/binary-tree-pruning.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode pruneTree(TreeNode root) {
        if (!hasOne(root)) {
            return null;
        }
        if (root != null) {
            root.left = pruneTree(root.left);
            root.right = pruneTree(root.right);
        }
        return root;
    }

    private boolean hasOne(TreeNode root) {
        if (root == null) {
            return false;
        }
        if (root.val == 1) {
            return true;
        }
        return hasOne(root.left) || hasOne(root.right);
    }
}

File: python/leetcode/binary-tree-pruning.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pruneTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if not self.hasOne(root):
            return None
        if (root is not None):
            root.left = self.pruneTree(root.left)
            root.right = self.pruneTree(root.right)
        return root

    def hasOne(self, root):
        if root is None:
            return False
        if root.val == 1:
            return True
        return self.hasOne(root.left) or self.hasOne(root.right)
